docs/init.rst

```diff
@@ -301,6 +301,21 @@ Arguably, you can abuse converters as one-argument validators:
    ValueError: invalid literal for int() with base 10: 'x'
 
 
+If a converter's first argument has a type annotation, that type will appear in the signature for ``__init__``.
+You can override this by passing the ``type`` argument or by giving an explicit type annotation.
+
+.. doctest::
+
+   >>> def str2int(x: str) -> int:
+   ...     return int(x)
+   >>> @attr.s
+   ... class C(object):
+   ...     x = attr.ib(converter=str2int)
+   ...     y = attr.ib(converter=str2int, type=int)
```

Comment:
I don't think this is correct behavior. The type parameter is equivalent to using a type annotation. IOW: it's what's stored on the class, _not_ what the initializer is accepting.

Reply:
That makes sense, and I've updated PR to reflect that. It's worth noting, though, that the `type` argument now becomes a no-op if a converter is present. I suppose that was already the case, though.

The only issues I see now are:
1. If `inspect.signature` finds that the converter doesn't take any arguments, an `IndexError` will be thrown at line 2208. In most cases, an error would have occurred anyway -- but it would have happened later, and it would have been more readable (e.g., `f() takes 0 positional arguments but 1 was given`).
2. I haven't special-cased `pipe`, because I don't see a good way to do that without modifying the definition of `pipe` (e.g., turning it into a callable class).

Reply:
2 - why don't just do `update_wrapper` in `pipe`? 
```
if not PY2 and converters:
    pipe_converter.__anno
    functools.update_wrapper(pipe_converter, converters[0])
```

Return-type annotation may be inaccurate, but the arguments annotations would anyway flow into `__init__`. To fix return ones, something like this can be done:
```
    if 'return' in converters[-1].__annotations__:
        pipe_converter.__annotations__['return'] = converters[-1].__annotations__['return']
    else:
        del converters[-1].__annotations__['return']
```

but I'm not sure if this approach is good enough.

Reply:
Hm the update_wrapper is intriguing, but the problem is that it also copies over `__name__` etc which would make it confusing.

But I guess we could do just a `pipe_converter.__annotations__ = converters[0].__annotations__` and then the handling of return you've outlined?

Reply:
You can just take only annotations on update_wrapper:
```
In [2]: def test_me(a: int, b: str): pass                                                                                                                                                                  

In [3]: def test_wrapper(*args, **kwargs): pass                                                                                                                                                            

In [4]: update_wrapper(test_wrapper, test_me, assigned=('__annotations__',))                                                                                                                                   
Out[4]: <function __main__.test_wrapper(a: int, b: str)>
```

Please, note that `update_wrapper`also copies signature information (which is important in this case, as annotations are parsed from signature) and just copying `__annotations__` wouldn't do the trick.

Reply:
Sure, why not then.